File: java/src/test/java/trinsic/AccountServiceTest.java
package trinsic;

import com.google.protobuf.InvalidProtocolBufferException;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import trinsic.okapi.DidException;
import trinsic.services.AccountService;
import trinsic.services.TrinsicService;
import trinsic.services.account.v1.AuthorizeWebhookRequest;
import trinsic.services.account.v1.LoginRequest;
import trinsic.services.account.v1.LoginResponse;

import java.util.concurrent.ExecutionException;

class AccountServiceTest {
    private static String myEcosystemIdOrName = "default";
  @Test
  public void testLogin() {
    var trinsic = new TrinsicService(TrinsicUtilities.getTrinsicServiceOptions());
    LoginResponse myLoginResponse = LoginResponse.newBuilder().build();
    try {

      // loginRequest() {
      var loginResponse =
          trinsic
              .account()
              .login(LoginRequest.newBuilder().setEmail("bob@example.com").setEcosystemId(myEcosystemIdOrName).build())
              .get();
      // }
    } catch (Exception ignored) {
      // This is expected
    }

    try {
      // loginConfirm() {
      var authToken = trinsic.account().loginConfirm(myLoginResponse.getChallenge(), "12345").get();
      // }

      // setAuthTokenSample() {
      trinsic.setAuthToken(authToken);
      // }
    } catch (Exception ignored) {
    }
  }

  @Test
  public void testAuthWebhook()
      throws ExecutionException, InterruptedException, InvalidProtocolBufferException,
          DidException {
    var trinsic = new TrinsicService(TrinsicUtilities.getTrinsicServiceOptions());

    var profile = trinsic.account().loginAnonymous(myEcosystemIdOrName).get();
    trinsic.setAuthToken(profile);

    // authorizeWebhook() {
    var authorizeResponse =
        trinsic
            .account()
            .authorizeWebhook(
                AuthorizeWebhookRequest.newBuilder()
                    .addEvents("*") // Authorize all events
                    .build())
            .get();
    // }
  }

  @Test
  public void testProtectUnprotectAccount()
      throws ExecutionException, InterruptedException, InvalidProtocolBufferException,
          DidException {
    // accountServiceConstructor() {
    var trinsic = new TrinsicService(TrinsicUtilities.getTrinsicServiceOptions());
    // }
    // accountServiceSignIn() {
    var myProfile = trinsic.account().signIn(myEcosystemIdOrName).get();
    // }

    // protectUnprotectProfile() {
    var code = "1234";
    var myProtectedProfile = AccountService.protect(myProfile, code);
    var myUnprotectedProfile = AccountService.unprotect(myProtectedProfile, code);
    // }

    Assertions.assertThrows(
        Exception.class,
        () -> {
          trinsic.setAuthToken(myProtectedProfile);
          Assertions.assertEquals(
              myProtectedProfile, trinsic.account().getOptionsBuilder().getAuthToken());
          trinsic.account().getInfo().get();
        });

    Assertions.assertDoesNotThrow(
        () -> {
          trinsic.setAuthToken(myUnprotectedProfile);
          // getInfo() {
          var info = trinsic.account().getInfo().get();
          // }
        });
  }
}

File: web/trinsic-web/test/AccountService.test.ts
import {
  AccountService,
  AuthorizeWebhookRequest,
  LoginRequest,
  TrinsicService,
} from "../src";

import { getTestServerOptions, setTestTimeout } from "./env";

const options = getTestServerOptions();

async function printGetInfo(service: TrinsicService, profile: string) {
  service.options.authToken = profile;
  const info = await service.account().getInfo();
}

describe("AccountService Unit Tests", () => {
  setTestTimeout();

  it("login / loginconfirm", async () => {
    let trinsic = new TrinsicService(options);

    // loginRequest() {
    const loginResponse = await trinsic.account().login(
      LoginRequest.fromPartial({
        email: "bob@example.com",
      })
    );
    //}

    // We expect this to fail; we have no real auth flow here
    try {
      // loginConfirm() {
      const authToken = await trinsic
        .account()
        .loginConfirm(loginResponse.challenge, "12345");
      //}

      // setAuthTokenSample() {
      trinsic.options.authToken = authToken;
      //}
    } catch {}
  });

  it("authorize webhook", async () => {
    let trinsic = new TrinsicService(options);
    let account = await trinsic.account().loginAnonymous();

    trinsic.options.authToken = account.toString(); //wat

    // authorizeWebhook() {
    const response = await trinsic.account().authorizeWebhook(
      AuthorizeWebhookRequest.fromPartial({
        events: ["*"], // Authorize all events
      })
    );
    //}
  });

  it("protect/unprotect account profile", async () => {
    let trinsic = new TrinsicService(options);
    let myProfile = await trinsic.account().loginAnonymous();
    await printGetInfo(trinsic, myProfile);

    const code = "1234";
    const myProtectedProfile = await AccountService.protect(myProfile, code);
    try {
      await printGetInfo(trinsic, myProtectedProfile);
      fail("previous line should have thrown.");
    } catch {}

    const myUnprotectedProfile = await AccountService.unprotect(
      myProtectedProfile,
      code
    );
    await printGetInfo(trinsic, myUnprotectedProfile);
  });
});
